### PR TITLE
Add online content library

### PR DESCRIPTION
**Pull request by rubenwardy**
Mainly posting this here to act as an issue tracker, and also get early feedback on design. This should be merged before the public release of the database website to allow for wrangling out bugs way before the 0.5.0 release

The backend is available at https://contentdb.rubenwardy.com
Note that it's currently in development and unstable, the database will be reset a few times before release.

- [x] Download package metadata
- [x] Package filtering by query
- [x] Add support for mods
- [x] Add package details page for more info
- [x] Pagination
- [x] Downloading/installing dialog
- [x] Add support for uninstalling mods
- [x] Update package path after installing or uninstalling
- [x] Add support for games
- [x] Add support for texture packs
- [x] Redo mainmenu tabs
- [x] Filter by package type
- [x] Make translatable
- [x] Fix memory leak in ModApiMainMenu::l_get_package_list
- [x] Detect texture pack installed

**Future PRs:**

- [ ] Add screenshot support
- [ ] Add package updating/upgrading
- [ ] Add tag support
- [ ] Get tags from JSON API

### CODE REVIEW

**Comment by rubenwardy** [comment]
> update support ?

See todo list

> CSM support ?

No.

**Comment by rubenwardy** [comment]
Okay, I just need to clean some stuff up. I'd like to leave the other features (screenshots, tags, and updating) to future PR(s) because this one is already pretty big

File: src/script/lua_api/l_mainmenu.cpp [hunk 1]
```diff
@@ -972,6 +981,71 @@ int ModApiMainMenu::l_get_screen_info(lua_State *L)
 	return 1;
 }
 
+int ModApiMainMenu::l_get_package_list(lua_State *L)
+{
+	std::string url = "";
+	try {
+		url = g_settings->get("contentdb_url");
+	} catch(SettingNotFoundException &e) {
+		lua_pushnil(L);
```
Comment (SmallJoker) [inline src/script/lua_api/l_mainmenu.cpp]: Does not happen as soon it's contained in `defaultsettings`

File: po/cs/minetest.po [hunk 2]
```diff
@@ -4833,8 +4833,8 @@ msgstr "cURL timeout"
 #~ msgid "Downloading"
 #~ msgstr "Stahuji"
 
-#~ msgid "Gamemgr: Unable to copy mod \"$1\" to game \"$2\""
-#~ msgstr "Gamemgr: Nepovedlo se zkopírovat mod \"$1\" do hry \"$2\""
+#~ msgid "pkgmgr: Unable to copy mod \"$1\" to game \"$2\""
```
[the remark this answers is not in the document]
Comment (rubenwardy) [reply po/cs/minetest.po]: yeah, I need to remove this

File: mods/mods_here.txt [hunk 3]
```diff
@@ -1,4 +0,0 @@
-You can install Minetest mods by copying (and extracting) them into this folder.
```
[the remark this answers is not in the document]
Comment (rubenwardy) [reply mods/mods_here.txt]: I tried readding this but it wouldn't add it, probably need a force add

**Comment by nerzhul** [comment]
why remove the namespace ? namespace is just well written C++. The name is not good, but it's nice to see some :)

**Comment by rubenwardy** [comment]
It's wrong to have it just for packages, I'd like to have it for all the content

**Comment by rubenwardy** [comment]
So, my preference for this is to have it merged and work on adding more features in future PRs. It doesn't matter that it's not feature complete yet

This'll make it much easier to find bugs before the release, by there being more time

File: builtin/mainmenu/pkgmgr.lua [hunk 4]
```diff
@@ -333,43 +369,74 @@ function modmgr.installmod(modfilename,basename)
...
-	modmgr.refresh_globals()
+--------------------------------------------------------------------------------
+function pkgmgr.install(type, modfilename, basename)
+	local archive_info = pkgmgr.identify_filetype(modfilename)
+	local path = pkgmgr.extract(archive_info)
 
+	if path == nil then
+		return nil,
+			fgettext("Install: file: \"$1\"", archive_info.name) ..
+			fgettext("\nInstall: unsupported filetype \"$1\" or broken archive",
```
Comment (SmallJoker) [inline builtin/mainmenu/pkgmgr.lua]: New line character should be moved outside the function.

File: builtin/mainmenu/tab_content.lua [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+
+		packages = filterlist.create(
+						function() return packages_raw end, --refresh
+						pkgmgr.compare_package, --compare
+						function(element, uid) --uid match
+							return (element.type == "game" and element.id == uid) or
+									element.name == uid
+						end,
+						nil, --filter
+						{})
```
Comment (SmallJoker) [inline builtin/mainmenu/tab_content.lua]: Possibly copy&pasted but please correct the indents.

File: builtin/mainmenu/tab_content.lua [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+
+
+	if fields.btn_mod_mgr_disable_txp then
+		core.settings:set("texture_path", "")
+		packages = nil
+		return true
+	end
+
+	if fields["mod_mgt_open_dlg_accepted"] ~= nil and
+		fields["mod_mgt_open_dlg_accepted"] ~= "" then
```
Comment (SmallJoker) [inline builtin/mainmenu/tab_content.lua]: Missing indent, in this case redundant `~= nil`
Line below: missing space after `,`

File: po/cs/minetest.po (hunk 2 above)
[the remark this answers is not in the document]
Comment (SmallJoker) [reply po/cs/minetest.po]: Still need to remove this.

File: src/content/packages.h [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+	}
+};
+
+#if USE_CURL
+std::vector<Package> getPackagesFromURL(const std::string &url);
+#else
+inline std::vector<Package> getPackagesFromURL(const std::string &url)
+{
+	return std::vector<Package>();
+}
```
Comment (SmallJoker) [inline src/content/packages.h]: Alternative non-curl method should be moved to the *.cpp file (will happen earlier or later anyway).

File: src/content/mods.cpp [hunk 8]
```diff
@@ -74,7 +77,9 @@ void parseModContents(ModSpec &spec)
 			mod_conf_has_depends = true;
 			std::string dep = info.get("depends");
 			dep.erase(std::remove_if(dep.begin(), dep.end(),
-					static_cast<int(*)(int)>(&std::isspace)), dep.end());
+						  static_cast<int (*)(int)>(
+								  &std::isspace)),
```
Comment (SmallJoker) [inline src/content/mods.cpp]: You may also use one tabulator as indent if visual space is an issue.
Wrong/Mixed indents.

File: src/content/mods.cpp [hunk 9]
```diff
@@ -84,7 +89,9 @@ void parseModContents(ModSpec &spec)
 			mod_conf_has_depends = true;
 			std::string dep = info.get("optional_depends");
 			dep.erase(std::remove_if(dep.begin(), dep.end(),
-					static_cast<int(*)(int)>(&std::isspace)), dep.end());
+						  static_cast<int (*)(int)>(
+								  &std::isspace)),
```
Comment (SmallJoker) [inline src/content/mods.cpp]: Ditto.

File: builtin/common/misc_helpers.lua [hunk 10]
```diff
@@ -551,6 +551,16 @@ function table.copy(t, seen)
 	end
 	return n
 end
+
+
+function table.insert_all(t, other)
```
Comment (SmallJoker) [inline builtin/common/misc_helpers.lua]: Needs documentation.

File: src/content/mods.cpp (hunk 8 above)
Comment (rubenwardy) [reply src/content/mods.cpp]: This was done by clang-format, please fix that instead

File: src/content/mods.cpp (hunk 9 above)
Comment (rubenwardy) [reply src/content/mods.cpp]: This was done by clang-format, please fix that instead

File: src/content/packages.h (hunk 7 above)
Comment (rubenwardy) [reply src/content/packages.h]: We don't support anything that isn't cURL for downloading

**Comment by twoelk** [comment]
I wonder if this could include some management tools like listing all installed mods and sorting them for example by last update or making a list of mods or even nodes used in a world.

File: src/content/packages.h (hunk 7 above)
Comment (SmallJoker) [reply src/content/packages.h]: About the client media can be downloaded without cURL - but I didn't think about that it's done using our own protocol, thus no HTTP/TCP use. These lines are fine.

**Comment by rubenwardy** [comment]
I'd like to have a CLI tool - especially for servers - but not in this PR

**Pull request merged**